jupyterlab_gcedetails/src/components/machine_type_select.tsx

```diff
@@ -0,0 +1,257 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from 'react';
+import { stylesheet } from 'typestyle';
+import List from '@material-ui/core/List';
+import ListItem from '@material-ui/core/ListItem';
+import ListItemText from '@material-ui/core/ListItemText';
+import Divider from '@material-ui/core/Divider';
+import ChevronRightRoundedIcon from '@material-ui/icons/ChevronRightRounded';
+import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
+import OutlinedInput from '@material-ui/core/OutlinedInput';
+import InputAdornment from '@material-ui/core/InputAdornment';
+import Grid from '@material-ui/core/Grid';
+import Paper from '@material-ui/core/Paper';
+import { Option, MACHINE_TYPES } from '../data';
+
+export const STYLES = stylesheet({
+  itemGroup: {
+    overflowY: 'auto',
+    height: '240px',
+    width: '233px',
+  },
+  menuContainer: {
+    paddingTop: '8px',
+    paddingBottom: '8px',
+    width: '468px',
+  },
+});
+
+/* Item component that the user can select on the menu */
+
+interface ItemProps {
+  option: Option;
+  onSelect: (o: Option) => void;
+  selected: boolean;
+}
+
+export function Item(props: ItemProps) {
+  const { option, onSelect, selected } = props;
+
+  return (
+    <ListItem onClick={e => onSelect(option)} button selected={selected}>
+      <ListItemText primary={option.value} secondary={option.text} />
+    </ListItem>
+  );
+}
+
+/* Header item component that, on hover, changes the options the user can select 
+from on the menu */
+
+interface HeaderItemProps {
+  label: string;
+  onHover: () => void;
+}
+
+interface HeaderItemState {
+  isHovering: boolean;
+}
+
+class HeaderItem extends React.Component<HeaderItemProps, HeaderItemState> {
+  constructor(props: HeaderItemProps) {
+    super(props);
+
+    this.state = {
+      isHovering: false,
+    };
+  }
+
+  render() {
+    const { label, onHover } = this.props;
+
+    return (
+      <div onMouseEnter={() => onHover()}>
+        <ListItem button>
+          <ListItemText primary={label} />
+          <ChevronRightRoundedIcon />
+        </ListItem>
+      </div>
+    );
+  }
+}
+
+/* Nested select body component */
+
+interface NestedSelectBodyProps {
+  onSelect: (o: Option) => void;
+  selectedOption: Option;
+}
+
+interface NestedSelectBodyState {
+  displayedOptions: Option[];
+}
+
+class NestedSelectBody extends React.Component<
+  NestedSelectBodyProps,
+  NestedSelectBodyState
+> {
+  constructor(props: NestedSelectBodyProps) {
+    super(props);
+
+    this.state = {
+      displayedOptions: MACHINE_TYPES[0].configurations,
```

Comment:
that's a good point, I was actually implementing this component with the fact that it may be valuable as a shared component in mind so the change should be easy enough - I'll add it in!